### PR TITLE
AbstractSchedules for scheduling output and diagnostics

### LINKED ISSUE #845: More general criteria for writing data, checkpointing, calculating diagnostics, printing progress statements, etc
We commonly need to specify criteria that triggers an action; eg saving of data. For example, the `OutputWriter`s have the keyword arguments `time_interval` (for saving after fixed intervals of time) and `iteration_interval` (for saving after a fixed number of iterations).

This method is a big clunky and doesn't allow for more general criteria that depend on, for example, flow features, wall clock time, etc. It'd be nice to have a general system for specifying `AbstractCriteria` that must be met for an action to occur. We can have types like `TimeInterval` and `IterationInterval`, or `WallTimeInterval`, for example, which will return true (given some argument like `model`) when saving / action needs to occur.

A similar problem exists for the printing of progress statements by `Simulation`, and also for calling "callback" functions during the time-stepping loop of `run!(simulation)`.

### LINKED ISSUE #853: Better criterion for writing output
Currently, output writers accept two types of criterion for writing output: a `time_interval`, and an `iteration_interval`. For example, a user might instantiate a `JLD2OutputWriter` with

```julia
field_outputs = FieldOutputs(merge(model.velocities, model.tracers))

simulation.output_writers[:fields] = JLD2OutputWriter(model, field_outputs, prefix="fields", time_interval=2minute)
```

which specifies output every 2 minutes of simulation time. `iteration_interval` is used similarly. These criteria are triggered by the function `time_to_run`:

https://github.com/CliMA/Oceananigans.jl/blob/994eae3d4550898914309e535ccbfd4655854646/src/Utils/output_writer_diagnostic_utils.jl#L55-L65

The function `time_interval_is_ripe` is, for example,

https://github.com/CliMA/Oceananigans.jl/blob/994eae3d4550898914309e535ccbfd4655854646/src/Utils/output_writer_diagnostic_utils.jl#L46-L53

This API is simple and effective for many purposes. However, it is not general or extensible. For example, a user cannot specify output to be saved based on other criteria, like some property of the flow, or on the elapsed wall time.

A more general method might use types. For example, we can define `AbstractCriterion` like `TimeInterval` and `IterationInterval`. The above example would become

```julia
field_outputs = FieldOutputs(merge(model.velocities, model.tracers))

simulation.output_writers[:fields] = JLD2OutputWriter(model, field_outputs, prefix="fields", write_criterion=TimeInterval(2minute))
```

By making the criteria callable objects, we could use syntax like `output_writer.write_criteria(model)` as a replacement for `time_to_run(clock, output_writer)`.

### LINKED ISSUE #1019: `JLD2OutputWriter` needs a nice show method
It makes the examples look kinda ugly and would generally be useful

image

### PR DESCRIPTION
**Pull request by glwagner**
This PR provides a substantial rewrite to the user API for "scheduling" output and diagnostics.

Previously, output was scheduled by specifying one of two keyword arguments when constructing the `AbstractOutputWriter`: `time_interval`, and `iteration_interval`. An example from the documentation is

```julia
simulation.output_writers[:field_writer] =
    NetCDFOutputWriter(model, fields, filename="output_fields.nc", time_interval=60)

# output
NetCDFOutputWriter (time_interval=60): output_fields.nc
├── dimensions: zC(16), zF(17), xC(16), yF(16), xF(16), yC(16), time(0)
└── 2 outputs: ["T", "u"]
```

Time-averaging was specified by providing one or two additional keyword arguments: `time_average_interval` and (optionally) `time_average_stride`.

This PR eliminates these keyword arguments in favor of a single argument `schedule`. In general, the `schedule` is an object which returns `true` (when output should be written, or a diagnostic calculated) and `false` (otherwise).

This PR implements four `AbstractSchedule`s, and more are possible:

* `TimeInterval`: periodic schedule that reoccurs on a interval of model time.
* `IterationInterval`: periodic schedule that reoccurs on an iteration of model iterations
* `WallClockInterval` periodic schedule that reoccurs on an interval of "wall time", as kept by the clock on your wall
* `AveragedTimeInterval`: periodic schedule that reoccurs on an interval of model time, and specifies time-averaging of output over a `window` (and a default `window` equal to `interval` is now provided for friendliness)

The old syntax thus becomes

```julia
simulation.output_writers[:field_writer] =
    NetCDFOutputWriter(model, fields, filepath="fields.nc", schedule=TimeInterval(60))
```

and averaging is specified (with a time window of 20 seconds) by writing

```julia
simulation.output_writers[:field_writer] =
    NetCDFOutputWriter(model, fields, filepath="fields.nc", schedule=AveragedTimeInterval(60, window=20))
```

This PR starts updating the examples and docs, though we still need to

- [x] finish updating verification scripts
- [x] finish updating docs

Since we now have a fairly convincing test about the accuracy of `WindowedTimeAverage`, I think we should also

- [x] add documentation for `WindowedTimeAverage`

which is apropos to this PR since it changes the syntax through which averaging is specified.

Eventually, we should use the "scheduling" concept for both callafter function in `run!(simulation)` and the `TimeStepWizard` (separately), which will get rid of the `iteration_interval` argument in `run!(simulation)`.

PS: should we use `WindowAveragedTimeInterval` rather than `AveragedTimeInterval`? I wasn't sure if the added verbosity added clarity here (the important part is the "Averaging") but I'm open to considering it.

Resolves #1019 
Resolves #853 
Resolves #845

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
> PS: should we use `WindowAveragedTimeInterval` rather than `AveragedTimeInterval`? I wasn't sure if the added verbosity added clarity here (the important part is the "Averaging") but I'm open to considering it.

I think `AveragedTimeInterval` is fine since `window` is a keyword argument. But I guess by not setting `window` the term suggests that no windowing will occur and the time average is accumulated every iteration?

File: docs/src/model_setup/output_writers.md [hunk 1]
```diff
@@ -1,130 +1,73 @@
...
+Model data can be saved to NetCDF files along with associated metadata. The NetCDF output writer is generally used by
+passing it a dictionary of (label, field) pairs and any indices for slicing if you don't want to save the full 3D field.
 
-# output
-NetCDFOutputWriter (iteration_interval=1): stuff.nc
-├── dimensions: zC(16), zF(17), xC(16), yF(16), xF(16), yC(16), time(0)
-└── 3 outputs: ["profile", "slice", "scalar"]
+```@docs
+NetCDFOutputWriter
 ```
```
Comment (ali-ramadhan) [inline docs/src/model_setup/output_writers.md:48]: :100:

File: docs/src/model_setup/output_writers.md [hunk 2]
```diff
@@ -1,130 +1,73 @@
...
-of the function will be saved to the JLD2 file. For example, to write out 3D fields for w and T and a horizontal average
-of T every 1 hour of simulation time to a file called `some_data.jld2`
-
-```julia
-using Oceananigans
-using Oceananigans.OutputWriters
-using Oceananigans.Utils: hour, minute
+JLD2 is a fast HDF5 compatible file format written in pure Julia.
+JLD2 files can be opened in julia and in Python with the [h5py](https://www.h5py.org/)
+package.
```
Comment (ali-ramadhan) [inline docs/src/model_setup/output_writers.md:54]:
```suggestion
JLD2 files can be opened in Julia with the [JLD2.jl](https://github.com/JuliaIO/JLD2.jl) package
and in Python with the [h5py](https://www.h5py.org/) package.
```

**Review by ali-ramadhan — APPROVED** [review]
This is awesome! Much cleaner interface with tons of uses.

For now I guess you can only have one schedule per output writer. I think this is fine: if users want to specify multiple schedules they can create multiple output writers (or just define a custom `schedule(model)`. I.e. we shouldn't worry about this.

Thanks for updating the documentation on output writing, it was quite out of date and it's not the most enjoyable work. Also nice work on refactoring the `NetCDFOutputWriter` through multiple disptach!

Super pedantic but `Oceananigans.Util` now defines `minutes` and `hours` so we can use e.g. `TimeInterval(2minutes)` instead of `TimeInterval(2minute)` but maybe this is bad as it could signal a proliferation of exported names .

**Comment by glwagner** [comment]
> I guess by not setting a `window` `AveragedTimeInterval` suggests that no windowing will occur and the time average is accumulated every iteration?

Right, that's default! I think it makes sense.

**Comment by glwagner** [comment]
> we can use e.g. TimeInterval(2minutes)

Is this for examples and docs? It does obviously read better; I'm not sure about the downside of name proliferation. Maybe we can stew on it and update in a future PR?

**Comment by ali-ramadhan** [comment]
Just updated a few remaining jldoctests so tests should pass soon.

**Comment by ali-ramadhan** [comment]
> Is this for examples and docs? It does obviously read better; I'm not sure about the downside of name proliferation. Maybe we can stew on it and update in a future PR?

I was thinking both, i.e. encourage their use to improve script readability. Yeah we can just update things as we go along.

I guess the downside is that new users may not immediately notice that these common words, e.g. `minutes`, are actually being exported by `Oceananigans.Utils`. Perhaps if they were exported by `Oceananigans.Units` that would be clearer. Also, newcomers to Julia may not realize that you can write things like `2π` or `3hours` but this should not stop us from using this nice language feature.

**Comment by glwagner** [comment]
> Perhaps if they were exported by a submodule like `Oceananigans.Units` that would be clearer.

very much agree.

**Comment by glwagner** [comment]
> For now I guess you can only have one schedule per output writer. I think this is fine: if users want to specify multiple schedules they can create multiple output writers (or just define a custom `schedule(model)`. I.e. we shouldn't worry about this.

Yeah, I think eventually or if we get lots of requests it might make sense to add some kind of `MixedSchedule` / `MultiSchedule` / `JointSchedule` / `CombinedSchedule` feature that combines multiple schedules (either with `&&` or `||`). Which should be trivial I think. We probably want to design a schedule that activates output based on a criteria diagnosed from the model state (and then either stays activated, or is continuously re-evaluated). Futuristic stuff I think!

**Pull request merged**